### DISCUSSION TITLE
Single strange behavior

### DISCUSSION THREAD

**Discussion by VBAndCs** [dotnet/roslyn#48466] [Q&A] [answered]
Try this in C#
```C#
            var x = 123456781.0F;
            Console.WriteLine(x); // 123456784
            var x = 123456785.0F;
            Console.WriteLine(x); // 123456784
```

Why are `123456781` and `123456785` converted to `123456784`?
In fact you can test this faster in VB, as the editor changes the numbers as soon as you leave the line!
```VB
Dim x As Single = 123456779.0F
```

`123456779` will be converted to `123456776` and `123456771` to `123456768`!
What does this mean, and why does it happen?
This doesn't seem a round operation to me, and there is no obvious reason to do any rounding here.

**Comment by VBAndCs** [comment]
Double also does such exotic things. For example: `12345678912345691.0` will be converted to `12345678912345692.0`

**Comment by CyrusNajmabadi** [comment]
https://en.wikipedia.org/wiki/Single-precision_floating-point_format

> All integers with 7 or fewer decimal digits, and any 2^n for a whole number −149 ≤ n ≤ 127, can be converted exactly into an IEEE 754 single-precision floating-point value. 

These are not integers with 7 or fewer digits.  Your integers have 9 digits and are not a power of 2.  So they cannot be converted exactly.  As such, you get a close value as per the specific "string to FP" algorithm used.

**Comment by CyrusNajmabadi** [comment, marked as answer]
I recommend reading https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html to better understand this space and the subtleties therein.

> This doesn't seem a round operation to me, and there is no obvious reason to do any rounding here.

The reason this happens is because those values cannot be represented exactly by either 32bit or 64bit floating point numbers.  You can use System.Decimal here.  However, that too has fixed amount of space, and there are values it cannot represent.  When deciding which data type to use for numeric values, it's necessary to understand their precision and range and how that will affect your domain.

**Comment by CyrusNajmabadi** [comment]
@VBAndCs You may want to consider posting your questions to gitter.im/csharplang or dicord.gg/csharp as they're better avenues for questions/answers.  :)
